src/neosophia/agents/base.py

```diff
@@ -31,230 +32,321 @@ class Agent:
-        for resource in resources:
-
-            # Only do it for databases we haven't saved in the yaml file yet
-            if resource not in self.resources:
-                print(f'Adding new resource {resource}...')
-                description = autils.get_database_description(resource)
-                resource_yaml = autils.process_for_yaml(resource, description)
-                self.resources[resource] = yaml.safe_load(resource_yaml)[0]
-
-        # Save resource descriptions to yaml file
-        autils.write_dict_to_yaml(
-            self.resources, 'resources', self.resources_file)
-
-        # Builds a dictionary containing the function name as the key and a
-        # tuple containing the callable function and the entire function code
-        # as a string as the value
-        self.function_dict = autils.build_function_dict_from_modules(modules)
-
-        # Generate a dictionary of {function name: Callable} from the functions
-        # available in the modules passed in
-        self.function_calls = {}
-        for func_name, (call, func_str) in self.function_dict.items():
-            self.function_calls[func_name] = call
+        self.workspace_dir = workspace_dir
+        self.system_prompt = system_prompt
 
         # Manually add the `extract_answer` function that's used at the end of
         # every interaction
-        self.function_calls['extract_answer'] = self.extract_answer
-        self.function_calls['system_exit'] = sys.exit
-
-        # Convert functions list to yaml format and save in tools.yaml
-        for func_name, (_, func_str) in self.function_dict.items():
-
-            # Only do it for functions we haven't saved in the yaml file yet
-            if func_name not in self.tools:
-                print(f'Adding new function {func_name}...')
-                function_yaml = autils.convert_function_str_to_yaml(
-                    func_str)
-                self.tools[func_name] = yaml.safe_load(function_yaml)[0]
-
-        # Save function descriptions to yaml file
-        autils.write_dict_to_yaml(self.tools, 'functions', self.tools_file)
+        self.tools['extract_answer'] = Tool(
+            name='extract_answer',
+            function_str=None,
+            description='Tool to extract an answer given a question and data',
+            call=self.extract_answer
+        )
+
+        self.tools['system_exit'] = Tool(
+            name='system_exit',
+            function_str=None,
+            description='Tool to exit the program',
+            call=sys.exit
+        )
+
+    def calculate_prompt_cost(self, prompt: str):
+        """ Function to calculate the input or output cost """
+        num_tokens = autils.count_tokens(prompt, self.model_info.name)
+        return {
+            'input': num_tokens * self.model_info.input_token_cost,
+            'output': num_tokens * self.model_info.output_token_cost
+        }
+
+    def _toggle_items(self, items_dict, base_prompt, command):
```

Comment:
As written, this doesn't use `self` and is always called with `self.variables` so you could get rid of `items_dict` as an arg. Or make it a function instead of method

Reply:
It calls `self.execute`, and also `toggle_resources` calls it with `self.resources` for the `items_dict`, not `self.variables`.

Reply:
Ah makes sense